extensions/ftp/src/main/java/org/mule/extension/ftp/internal/AbstractFtpCopyDelegate.java

```diff
@@ -0,0 +1,156 @@
+    {
+        ConnectionHandler<FtpFileSystem> writerConnectionHandler;
+        final FtpFileSystem writerConnection;
+        try
+        {
+            writerConnectionHandler = getWriterConnection();
+            writerConnection = writerConnectionHandler.getConnection();
+        }
+        catch (ConnectionException e)
+        {
+            throw command.exception(String.format("FTP Copy operations require the use of two FTP connections. An exception was found trying to obtain second connection to" +
+                                                  "copy the path '%s' to '%s'", source.getPath(), targetPath), e);
+        }
+        try
+        {
+            if (source.isDirectory())
+            {
+                copyDirectory(Paths.get(source.getPath()), targetPath, overwrite, writerConnection, event);
+            }
+            else
+            {
+                copyFile(source, targetPath, overwrite, writerConnection, event);
+            }
+        }
+        catch (Exception e)
+        {
+            throw command.exception(format("Found exception copying file '%s' to '%s'", source, targetPath), e);
+        }
+        finally
+        {
+            writerConnectionHandler.release();
+        }
+    }
+
+    /**
+     * Performs a recursive copy of a directory
+     *
+     * @param sourcePath       the path to the directory to be copied
+     * @param target           the target path
+     * @param overwrite        whether to overwrite the target files if they already exists
+     * @param writerConnection the {@link FtpFileSystem} which connects to the target endpoint
+     * @param event            the {@link MuleEvent} which triggered this operation
+     */
+    protected abstract void copyDirectory(Path sourcePath, Path target, boolean overwrite, FtpFileSystem writerConnection, MuleEvent event);
+
+    /**
+     * Copies one individual file
+     *
+     * @param source           the {@link FileAttributes} for the file to be copied
+     * @param target           the target path
+     * @param overwrite        whether to overwrite the target files if they already exists
+     * @param writerConnection the {@link FtpFileSystem} which connects to the target endpoint
+     * @param event            the {@link MuleEvent} which triggered this operation
+     */
+    protected void copyFile(FileAttributes source, Path target, boolean overwrite, FtpFileSystem writerConnection, MuleEvent event)
+    {
+        FileAttributes targetFile = command.getFile(target.toString());
+        if (targetFile != null)
+        {
+            if (overwrite)
+            {
+                fileSystem.delete(targetFile.getPath());
+            }
+            else
+            {
+                throw command.exception(String.format("Cannot copy file '%s' to path '%s' because it already exists", source.getPath(), target));
+            }
+        }
+
+        try (InputStream inputStream = fileSystem.retrieveFileContent(source))
```

Comment:
Could it be the case that the InputStream is already in the payload in this scenario?